File: pi-extension/errors.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import { ForgeError, forgeErrorEnvelope } from "./errors.ts";

test("error envelopes remove repeated code prefixes and redact private fields", () => {
  const envelope = forgeErrorEnvelope(new ForgeError("claim_denied", "claim_denied: another execution", 409, false, {
    hint: "Bearer private-token",
    data: { cleanup_pending: false, execution_token: "private-token", attempt_id: "private-attempt" },
  }));
  assert.deepEqual(envelope, {
    status: 409,
    error: { code: "claim_denied", message: "another execution", ambiguous: false, hint: "Bearer [REDACTED]", data: { cleanup_pending: false } },
  });
});

File: cmd/forged/errors_test.go
package main

import (
	"errors"
	"testing"

	"github.com/rkbkosp/collaboration-forge/internal/checkout"
)

func TestForgedErrorPreservesStructuredCheckoutFieldsSafely(t *testing.T) {
	err := &checkout.Error{
		Status:    503,
		Code:      "claim_denied",
		Message:   "lease is held",
		Hint:      "wait before retrying",
		Data:      map[string]any{"cleanup_pending": false, "execution_token": "private"},
		Ambiguous: true,
	}
	converted := forgedErrorFrom(err)
	if converted.code != "claim_denied" || converted.status != 503 || converted.hint != "wait before retrying" || !converted.ambiguous {
		t.Fatalf("structured fields lost: %+v", converted)
	}
	if _, ok := converted.data["execution_token"]; ok {
		t.Fatal("execution token escaped checkout error")
	}
}

func TestForgedErrorDoesNotExposeNativeCause(t *testing.T) {
	converted := forgedErrorFrom(errors.New("open /private/token-file: private-token"))
	if converted.code != "forged_error" || converted.message == "" || converted.message == "open /private/token-file: private-token" {
		t.Fatalf("native cause exposed: %+v", converted)
	}
}
